proposed implementation for security logging, needs review.

preichel-cg opened the pull request:
#569 
- new class SecureLogging which provides Markers for case-specific logging.
- new optional dependency org.owasp which extends functionality of Markers.
- corresponding appenders for console and file with example filter implementations.
- MaskingConverter for confidential information, until OWASP issue is fixed.

hohwille commented on modules/logging/src/main/java/io/oasp/module/logging/common/impl/SecureLogging.java:
```diff
@@ -0,0 +1,242 @@
+  /**
+   * @param extClass new value of {@link #getExtClass}.
+   */
+  public static void setExtClass(String extClass) {
```
Seems over engineered: getting and setting of "extClass" and "methodName" is IMHO a flexibility that none will actually need. Even worse I assume that it might introduce bugs with bootstrapping when the setter is called too late in the startup process. In the end the arguments of the getMarker method are also hard-coded. For KISS I would propose to just change these two attributes to constants and remove these getter/setter pairs.

hohwille commented on modules/logging/src/main/java/io/oasp/module/logging/common/impl/SecureLogging.java:
```diff
@@ -0,0 +1,242 @@
+ *
+ * <pre>
+ * {@code
+ * LOG.info(SecureLogging.SECURITY_FAILURE_CONFIDENTIAL, "Confidential Security Failure message.");
```
Nice to have but I would propose to use {@link SecureLogging},{@link #SECURITY_FAILURE_CONFIDENTIAL} etc.
This makes the JavaDoc even more specific and expressive, ensures there are no typos and is refactoring safe.

hohwille commented on modules/logging/src/main/java/io/oasp/module/logging/common/impl/SecureLogging.java:
```diff
@@ -0,0 +1,242 @@
+  /**
+   * Marker for Top Secret log events.
+   */
+  public static final Marker TOP_SECRET = MarkerFactory.getDetachedMarker(TOP_SECRET_MARKER_NAME);
```
We do not have a clear usage concept yet. However, if something is TOP_SECRET then the most secure solution is to never let it reach any logger at all. If something is present here it will be used some day (by someone who does not question this). Therefore I personally consider this Marker as misleading and would remove it.

We should in parallel open a discussion and try to collect experience in the area of Markers and esp. OWASP security markers. To finalize the entire initiative it would be great to have a more clear usage concept and some examples in the sample app (my-thai-star) in the future. Of course this last point is not a requirement for you to pick up before we can accept this PR.

hohwille commented on modules/logging/src/main/java/io/oasp/module/logging/common/impl/SecureLogging.java:
```diff
@@ -0,0 +1,242 @@
+ * }
+ * </pre>
+ */
+public class SecureLogging {
```
Could you add a simple JUnit Test that at least verifies that logging with such marker actually works at all (classloading, etc. is sane).

hohwille commented on modules/logging/pom.xml:
```diff
@@ -24,6 +24,11 @@
     <dependency>
+      <groupId>org.owasp</groupId>
+      <artifactId>security-logging-logback</artifactId>
+      <version>1.1.3</version>
```
move versions to our BOM: https://github.com/oasp/oasp4j/blob/develop/bom/pom.xml

hohwille commented on modules/logging/pom.xml:
```diff
@@ -24,6 +24,11 @@
     </dependency>
     <dependency>
+      <groupId>org.owasp</groupId>
+      <artifactId>security-logging-logback</artifactId>
```
Please declare this dependency as optional.
You may add it (non optional) in current sample app. People who do not want it can then remove it from there.
But logging module should not force you to use logback nor OWASP stuff.

preichel-cg commented:
@hohwille Hi Jörg, I have addressed all of your mentioned points and made 2 commits, which now show up in this PR.
- I added a test class for SecureLogging, which checks logging, filtering and masking when using logback.
- Unfortunately I did not yet succeed to create one when using log4j instead...

hohwille commented on modules/logging/src/test/java/io/oasp/module/logging/common/impl/SecureLoggingLogbackTest.java:
```diff
@@ -0,0 +1,248 @@
+
+    // Check the message being logged is reasonable
+    String layoutMessage = this.encoder.getLayout().doLayout(loggingEvent);
+    System.out.println("test(): layoutMessage = " + layoutMessage);
```
remove

hohwille commented on modules/logging/src/test/java/io/oasp/module/logging/common/impl/SecureLoggingLogbackTest.java:
```diff
@@ -0,0 +1,248 @@
+
+    // Check the message being logged is reasonable
+    String layoutMessage = this.encoder.getLayout().doLayout(loggingEvent);
+    System.out.println("test(): layoutMessage = " + layoutMessage);
```
remove

hohwille commented on modules/logging/src/test/java/io/oasp/module/logging/common/impl/SecureLoggingLogbackTest.java:
```diff
@@ -0,0 +1,248 @@
+  public void testInitWithMultiMarkerClass() {
+
+    // skip test if the dependency is not available.
+    if (!SecureLogging.hasExtClass())
```
never omit curly braces in Java. Your IDE does not seem to be setup properly otherwise you would get warnings.

hohwille commented on modules/logging/src/test/java/io/oasp/module/logging/common/impl/SecureLoggingLogbackTest.java:
```diff
@@ -0,0 +1,248 @@
+    // SecureLogging.initMarkers() is called by the loc below:
+    Marker multiMarker = SecureLogging.SECURITY_SUCCESS_CONFIDENTIAL;
+    Marker securMarker = SecureLogging.SECURITY_SUCCESS;
+    Marker confidMarker = SecureLogging.CONFIDENTIAL;
```
IMHO making these to local variables is pointless and even contradictory for readability of asserts.

hohwille commented on modules/logging/src/test/java/io/oasp/module/logging/common/impl/SecureLoggingLogbackTest.java:
```diff
@@ -0,0 +1,248 @@
+
+    // Check the message being logged is reasonable
+    String layoutMessage = this.encoder.getLayout().doLayout(loggingEvent);
+    assertFalse("formatted log message must not be empty.", layoutMessage.isEmpty());
```
Applies for all asserts. Even though AssertJ is no big advantage here we are heavily working on educating people to stop using default JUnit asserts or hamcrest. Therefore OASP should follow best practices and avoid using this.

hohwille commented on modules/logging/src/test/java/io/oasp/module/logging/common/impl/SecureLoggingLogbackTest.java:
```diff
@@ -0,0 +1,248 @@
+   * Test if logging works at all, without using any Markers.
+   */
+  @Test
+  public void testDefaultLogEvent() {
```
best practice for such kind of simple tests is to follow this pattern:
```
// given
Type data = ...
// when
invokeMethodToTest(data);
// then
assertThat(...)...
```
See https://github.com/oasp/oasp4j/wiki/guide-testing#general-best-practices

hohwille commented on modules/logging/src/test/java/io/oasp/module/logging/common/impl/SecureLoggingLogbackTest.java:
```diff
@@ -38,7 +39,10 @@
 @RunWith(MockitoJUnitRunner.class)
-public class SecureLoggingLogbackTest extends ModuleTest {
+@Category(CategoryModuleTest.class)
+public class SecureLoggingLogbackTest {
```
You should still use ModuleTest but need to override `doSetup()` then.

preichel-cg commented:
Hi Jörg,
the new main changes are:

- SecureLoggingLogbackTest extends ModuleTest
- use AssertJ
- use `//given` - `//when` - `//then`
- rename static variables
- change logger class and add casts logback logger in doSetUp/doTearDown.

open points:
- your remark that local Marker variables are contradictory: for me personally they improve readability (e.g. `assertThat` statements are much shorter) and reduce the chance of making inconsistent changes. And especially with the `//given` - `//when` - `//then` logic they make clearer, that their correct initialization is a particular feature which shall be tested. But I will change it if you insist ;-).

Cheers,
Patrick

hohwille commented:
Still no concerns. Merging now.
